cancer_ai/rewarder.py

```diff
@@ -0,0 +1,60 @@
+from pydantic import BaseModel
+from datetime import datetime
+
+class CompetitionLeader(BaseModel):
+    hotkey: str
+    leader_since: datetime
+
+class RewarderConfig(BaseModel):
+    competition_leader_mapping: dict[str, CompetitionLeader]
+    scores: dict[str, float] # hotkey -> score
+
+class Rewarder():
+    def __init__(self, config: RewarderConfig):
+        self.competition_leader_mapping = config.competition_leader_mapping
+        self.scores = config.scores
+
+    def calculate_score_for_winner(self, competition_id: str, hotkey: str) -> tuple[float, float]:
+        num_competitions = len(self.competition_leader_mapping)
+        base_share = 1/num_competitions
+
+        # check if current hotkey is already a leader
+        if self.competition_leader_mapping[competition_id].hotkey == hotkey:
+            days_as_leader = (datetime.now() - self.competition_leader_mapping[competition_id].leader_since).days
+        else:
+            days_as_leader = 0
+            self.competition_leader_mapping[competition_id] = CompetitionLeader(hotkey=hotkey,
+                                                                           leader_since=datetime.now())
+
+        if days_as_leader > 14:
+            periods = (days_as_leader - 14) // 7
+            reduction_factor = max(0.1, 1 - 0.1 * periods)
+            final_share = base_share * reduction_factor
+            reduced_share = base_share - final_share
+            return final_share, reduced_share
+        return base_share, 0
+    
+    def update_scores(self):
+        num_competitions = len(self.competition_leader_mapping)
```

Comment:
we need to make sure, when deserializing state of miners on run, should check also against how many competitions are there, and if there is new one, it should create empty mapping. So also update RewardConfig to allow empty field under mapping